src/main/kotlin/com/terraformation/backend/ask/ChatService.kt

```diff
@@ -0,0 +1,126 @@
+      """
+      
+      Context information is below, surrounded by ---------------------
+
+      ---------------------
+      {question_answer_context}
+      ---------------------
+
+      Given the context and provided history information, reply to the user comment. If the answer
+      is not in the context, inform the user that you can't answer the question. You may use general
+      knowledge of geography and science to interpret the context. If there are multiple entries in
+      the context with similar information, prefer the newer ones if you can determine their dates.
+      """
+          .trimIndent()
+
+  private val includeVariablesAndDocumentsPrompt =
+      """
+      
+      Include a list of the submission document names and a list of the variable names of any
+      documents or variables that support the answer.
+      """
+          .trimIndent()
+
+  /**
+   * Logs requests and responses so you can see what exact prompts we're sending. To see the log
+   * messages, set `logging.level.org.springframework.ai.chat.client.advisor.SimpleLoggerAdvisor` to
+   * `DEBUG` in the application properties.
+   */
+  private val loggerAdvisor =
+      SimpleLoggerAdvisor(
+          { request -> request.toPrompt().contents },
+          { response -> response.result.output.text },
+          100)
+
+  /**
+   * Asks an LLM a question and returns the answer. Automatically includes contextual information in
+   * the prompt based on the project ID and the previous questions and answers in the conversation.
+   *
+   * @param projectId If non-null, only this project's information will be included as context.
+   *   Otherwise context may be pulled from all projects.
+   * @param conversationId If non-null, include previous messages from the conversation in the
+   *   prompt so the LLM knows what any followup questions are referring to. The current question
+   *   and answer will also be added to the conversation's history.
+   * @param showVariables If true, ask the LLM to say which variables and/or uploaded documents it
+   *   referenced when generating the answer.
+   */
+  fun askQuestion(
+      projectId: ProjectId?,
+      question: String,
+      conversationId: String? = null,
+      showVariables: Boolean = false
+  ): String? {
+    val template =
+        listOfNotNull(
+                basePromptTemplate,
+                if (showVariables) includeVariablesAndDocumentsPrompt else null,
+            )
+            .joinToString("\n")
+
+    val questionAnswerAdvisor =
+        QuestionAnswerAdvisor(
+            vectorStore,
+            SearchRequest.builder()
+                .apply {
+                  if (projectId != null) {
+                    filterExpression(FilterExpressionBuilder().eq("projectId", projectId).build())
+                  }
+                }
+                .topK(contextItemsToInclude)
+                .build(),
+            template)
+
+    return chatClientBuilder
+        .build()
```

Comment:
Would we eventually want this to not be built for every question? Or does it need to be to get all the context for the new question?

Reply:
It's possible we could share an instance, yeah. When I wrote this, I didn't want to have to worry about whether or not it's stateful in a way that would cause problems if there are multiple conversations going on concurrently. But maybe that's not an issue.